### ISSUE TITLE
Import trigger after successful export

### ISSUE THREAD

**Issue by tuegeb** [Flagbit/Magento-FACTFinder#135]
After the successful export of the product feed - optionally including stock and price feed - the import should be started automatically. This includes the data, suggest and recommendation import on FACT-Finder side. For this purpose you can use the built-in `ImportAdapter` of the FACT-Finder PHP library.
The import trigger should be implemented in the Magento cronjob implementation and the shell script.

**Comment by xpoback** [comment]
I couldn't find a way to trigger stock and price import. Any suggestions?

**Comment by tuegeb** [comment]
All configured file types are downloaded by FACT-Finder automatically, when using the additional parameter `download=true` in the data import call. By the way I cannot confirm that the Magento cronjob is triggered.

**Comment by tuegeb** [comment]
Open tasks:

- trigger the update of the store, price and finally product export before the data, suggest and recommendation import ist triggered
- use the specific configuration to ensure the correct `primaryChannel` is used for every store
- shell script should provide option exportAllTypesAndImport per store and should support this process globally with the option exportAllTypesAndImportForAllStores (= behaviour of the cronjob export option) for all the stores with configured ´primaryChannel`

**Comment by tuegeb** [comment]
@xpoback: Will this improvment be finalized soon?

**Comment by xpoback** [comment]
@tuegeb yes, in one or two days. Thanks for reminding.

**Comment by xpoback** [comment]
@tuegeb What is meant by `specific configuration` to ensure the correct primaryChannel is used for every store ?

**Comment by tuegeb** [comment]
The store specific configuration.

**Comment by tuegeb** [comment]
@xpoback: Are there any news regarding this issue?

**Comment by xpoback** [comment]
@tuegeb should Suggest and Recommendation import be triggered 90 late only for FF >= 7.1 or for all versions?

**Comment by tuegeb** [comment]
The delay should always be considered. The export + import sequence would be as follows:
- stock export creation
- price export creation
- product export creation
- if FTP upload is enabled, ZIP creation and upload
- data import for primary channel with download=true if FTP upload is disabled or download=false if FTP is used
- delay
- recommendation import for primary channel 
- suggest import for primary channel

**Comment by xpoback** [comment]
@tuegeb Regarding `exportAllTypesAndImport` and `exportAllTypesAndImportForAllStores` I think that the `AndImport` part is redundant, since the import is triggered anyways if it has been enabled in the configuration.

**Comment by tuegeb** [comment]
@xpoback: Yes, you are right. So only the options `exportAllTypesStore` and `exportAllTypes` (equal to the Magento Cronjob option in the global configuration scope) are needed. They will ensure that the export files are created in the correct sequence for all the stores that have a configured `primaryChannel`. The subsequent import requests will be made only, if the trigger option is enabled in the store specific (fallback website or default) settings.

**Comment by xpoback** [comment]
@tuegeb Should be fine now, please test.

**Comment by tuegeb** [comment]
@xpoback: Many thanks.
I have made some minor changes to the help message of the shell script.
When using the option `exportAllTypesForAllStores` the price and stock export will only be updated for the first store. For the second and further channels, only the product export will be updated.
In system.log I found the following errors:
```

2016-06-22T18:14:00+00:00 ERR (3): Warning: session_start(): Cannot send session cookie - headers already sent by (output started at magento\shell\factfinder.php:74)  in magento\app\code\core\Mage\Core\Model\Session\Abstract\Varien.php on line 133
2016-06-22T18:14:00+00:00 ERR (3): Warning: session_start(): Cannot send session cache limiter - headers already sent (output started at magento\shell\factfinder.php:74)  in magento\app\code\core\Mage\Core\Model\Session\Abstract\Varien.php on line 133
```
Could you check that, please?

**Comment by xpoback** [comment]
@tuegeb The logic why we update stock and price only for first store is not clear to me, but whatever.
Regarding the warning, during the export magento tries to start session somewhere in the product export model, but since we output messages to the console, so session cookie cannot be sent.
I can try to find out where exactly it happens, but I'm pretty sure it's somewhere in the core, since we don't do anything with session in the module.
The easiest way would be to just silence it in the shell script: session plays no role for us when using php-cli.

**Comment by tuegeb** [comment]
The problem is not the update order. 
The problem here is, that only the files for one store are completly updated.

**Comment by xpoback** [comment]
@tuegeb Strange, since on my installation all price and stock files were created\updated.

**Comment by tuegeb** [comment]
@xpoback: I could reproduce this incomplete price/stock files issue on Windows (XAMPP environment), but not on Linux anymore.
But there is still another issue: The delayed recommendation and suggest imports are started at the same time. Please ensure that they are not triggered at the same time. The suggest import should be triggered after the recommendation import has finished (The edge-case: recommendations trigger off, has to be considered). And probably just a minor issue: On Windows the delayed import isn't supported because of the undefined function `pcntl_fork`

**Comment by xpoback** [comment]
@tuegeb Then I suppose we start suggest import 180 seconds after the product one.
Are there people who run production magento shops on Windows? If so, we start imports immediately if the `pcntl_fork` is undefined, or is it a bad idea?

**Comment by tuegeb** [comment]
@xpoback: Okay, 180 s is fine (if recommendation import is enabled, otherwise just 90 s). But as this will not be needed in future version of FACT-Finder: Could you add an option - delayed import on/off (default: true) - to the settings, please?
Additionally there should be an OS check, to trigger the delayed import only on Linux environments and on other environments an immediate import should be made.

**Comment by xpoback** [comment]
@tuegeb things accomplished:
 - stock and price now updated for all stores
 - imports are triggered immediately if `pcntl_fork` is not supported (Windows)
 - each import type is triggered 90 seconds after the previous one

Please, check and confirm.

**Comment by tuegeb** [comment]
@xpoback: Many thanks. 
It turned out, that there is a similar issue like  #125. Could you make sure, that the store specific connection settings are used for the import trigger, please?
Thank you in advance.

**Comment by tuegeb** [comment]
@xpoback: Could you solve this issue in version 4.1.9, please? Thank you in advance.

**Comment by xpoback** [comment]
@tuegeb I checked triggerring the import for different store views and websites and always the correct primary channel was triggered. Export files were correct as well. 
The only thing is notable: stock and price files are identical within stores of the same website, since price and stock have website scope.
Please, provide more details so I could be able to reproduce the issue.

**Comment by tuegeb** [comment]
@xpoback:
The primary channels are totally fine. But if you have differing domain, context other settings in the different store configurations, always the one from the default configuration are used in the ImportAdapter.

**Comment by xpoback** [comment]
@tuegeb Added fix for that. This goes to the release 4.1.9

**Comment by tuegeb** [comment]
@xpoback: The fix d9f4b74 is missing in case of the suggest and recommendation import request.
